oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resource/BearerTokenErrorCodes.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.resource;
+
+/**
+ * Standard error codes defined by the OAuth 2.0 Authorization Framework: Bearer Token Usage.
+ *
+ * @author Vedran Pavic
+ * @since 5.1
+ * @see <a href="https://tools.ietf.org/html/rfc6750#section-3.1" target="_blank">RFC 6750 Section 3.1: Error Codes</a>
+ */
+public interface BearerTokenErrorCodes {
```

Comment:
I don't believe that is a good idea. This is a different set of error codes (see [RFC 6750 Section 3.1](https://tools.ietf.org/html/rfc6750#section-3.1)) to one defined in [RFC 6749 Section 4.1.2.1](https://tools.ietf.org/html/rfc6749#section-4.1.2.1) and they should be kept separately.

By extension, this also applies to your remarks on use of `OAuth2Error` and `OAuth2AuthenticationException`.

Reply:
The `invalid_token` and `insufficient_scope` error codes (new) are still *OAuth 2.0 Errors* and are simply registered in the [OAuth Extensions Error Registration](https://tools.ietf.org/html/rfc6750#section-6.2) as defined in [11.4.  OAuth Extensions Error Registry](https://tools.ietf.org/html/rfc6749#section-11.4).

`invalid_request` or  `OAuth2ErrorCodes.INVALID_REQUEST` can be leveraged for Resource Server as well rather than duplicating this error code for Resource Server.

Reply:
Joe, does being registered in the OAuth Extensions Error Registration promote those error codes outside of the context in which they were defined? For example, it seems to me that it would be inappropriate to use `insufficient_scope` outside of a Resource Server context. (correct me if I'm wrong here)

This would be my only concern with adding bearer-token-focused codes to an umbrella class, that being misuse in other scenarios.

What are your thoughts on maybe having `BearerTokenErrorCodes extends OAuth2ErrorCodes` and `BearerTokenError extends OAuth2Error`? 

As for simply using `OAuth2Error`, the error details for `insufficient_scope` say

> The request requires higher privileges than provided by the
         access token.  The resource server SHOULD respond with the HTTP
         403 (Forbidden) status code and MAY include the "scope"
         attribute with the scope necessary to access the protected
         resource.

meaning that `BearerTokenError` needs one more member variable than `OAuth2Error`.

Reply:
Regardless of the fact OAuth Extensions Error Registration, they are still scoped by their location of use and shouldn't be used interchangeably everywhere. By stuffing everything to `OAuth2ErrorCodes` (and reusing `OAuth2Error` and `OAuth2AuthenticationException`) we exactly encourage that.

IMO Nimubus SDK gets this right by having `com.nimbusds.oauth2.sdk.OAuth2Error` and `com.nimbusds.oauth2.sdk.token.BearerTokenError`.

Besides, we talked the other day how parts of Resource Server (namely filter and everything bearer token resolution related) has the potential of being used outside of OAuth2 flow, i.e. with bearer token of origin other than OAuth2. We can forget that if we start coupling this to core OAuth2 related concerns.

Reply:
I was anticipating the use of `invalid_token` and `insufficient_scope` outside of the Resource Server context but the more I thought about it the more I couldn't find a use case. So agree that we should not add these to `OAuth2ErrorCodes`.

I still feel that `BearerTokenAuthenticationException` and `BearerTokenError` are redundant. `BearerTokenAuthenticationException` is the same as `OAuth2AuthenticationException`. However, `OAuth2Error` is very similar to `BearerTokenError` but it adds that extra member `scope`. Do we really need this extra member? Why can't the `scope(s)` be specified in the `message` of `OAuth2AuthenticationException` when constructed? The way I see it happening is that the `AuthenticationProvider` or `AccessDecisionVoter` will ultimately create an `OAuth2AuthenticationException` with the following `OAuth2Error`:

`oauth2Error.errorCode` = "insufficient_scope"
`oauth2Error.description` = "The following scopes are required to access this resource -> scope1, scope2"

Reply:
Just to point out as well, `OAuth2Error` was specifically placed in `oauth2-core` and meant to be reused across the 3 roles/modules. It represents an OAuth 2.0 Error regardless of it's context. However, agree that context specific `ErrorCodes` should be placed in it's appropriate module.

Reply:
The exception specific to Bearer Token authentication is needed to ensure the accompanying implementation of `AuthenticationEntryPoint` (which is the subject of `spring-projects/spring-security#5125` that depends on what we're doing here) is guaranteed to return `WWW-Authenticate` header with only error codes allowed by the spec in https://tools.ietf.org/html/rfc6750#section-3.1.

If you'd use generic `OAuth2AuthenticationException` how do you guard from `WWW-Authenticate` being populated with error codes that shouldn't be used in this context?

Reply:
The `AuthenticationEntryPoint` implementation would accept an `OAuth2AuthenticationException` and it's associated `OAuth2Error`. One of the evaluations in the implementation would be:

```
if OAuth2Error.errorCode.equals("invalid_token")
... then send back 401 and WWW-Authenticate...
```

I don't see any issues here. If you still do, please be specific on the use case you are pointing out so I can answer it more effectively.

Reply:
@vpavic As discussed, let's keep `BearerTokenAuthenticationException` but would like you to consider `BearerTokenError` to `extend OAuth2Error` to leverage the 3 common attributes.

Reply:
Now that I got back to this, I recalled why I haven't extended the `OAuth2Error` initially - it's marked as `final`.

Reply:
I'll remove the `final` modifier and commit/push so you can leverage it.

Reply:
Thanks, sounds good.